azure/datalake/store/transfer.py

```diff
@@ -240,9 +245,14 @@ def __init__(self, adlfs, transfer, merge=None, nthreads=None,
         self._chunked = chunked
         self._unique_temporary = unique_temporary
         self._unique_str = uuid.uuid4().hex
+        self._progress_callback=progress_callback
+        self._progress_lock = threading.Lock()
```

Comment:
If you're willing to trade some memory for lower lock-contention then you can heavily scale back the use of the lock. `list.append()` is atomic, so if you append to a list the total bytes sent then you don't need to lock. And calling `sum()` is also atomic so you could do `sum(sent_chunks)`. And if memory is an issue you could then do a cheap check for when the buffer gets to a certain size and then collapse it. IOW:
```python
self._send progress_buffer = []
...
self._send_progress_buffer.append(length)  # Atomic
if len(self._send_progress_buffer) > 1000:  # Choose some number you're comfortable with.
    with self._progress_lock:
        # In-place replacement to keep non-locking guarantees.
        self._send_progress_buffer[:] = [sum(self._send_progress_buffer)]
# Assignment in this order guarantees that total >= sent if total increases dynamically.
sent = sum(self._send_progress)
total = self._progress_total_bytes
self._progress_callback(sent, total)
```
This example will never lose data, guaranteed (Python unofficial memory model is, "order isn't guaranteed, but you won't lose data"). And if self._progress_total_bytes changes its total as time progresses you at worst might end up slightly out of step with bytes sent versus the total at the time of the send, but as long as the total is dealt with first you shouldn't ever end up with more bytes "sent" than there are in total.